src/_pytest/assertion/rewrite.py

```diff
@@ -118,22 +117,19 @@ def exec_module(self, module):
         if write:
             try:
                 os.mkdir(cache_dir)
-            except OSError:
-                e = sys.exc_info()[1].errno
-                if e == errno.EEXIST:
-                    # Either the __pycache__ directory already exists (the
-                    # common case) or it's blocked by a non-dir node. In the
-                    # latter case, we'll ignore it in _write_pyc.
-                    pass
-                elif e in {errno.ENOENT, errno.ENOTDIR}:
-                    # One of the path components was not a directory, likely
-                    # because we're in a zip file.
-                    write = False
-                elif e in {errno.EACCES, errno.EROFS, errno.EPERM}:
-                    state.trace("read only directory: %r" % os.path.dirname(fn))
-                    write = False
-                else:
-                    raise
+            except FileExistsError:
+                # Either the __pycache__ directory already exists (the
+                # common case) or it's blocked by a non-dir node. In the
+                # latter case, we'll ignore it in _write_pyc.
+                pass
+            except (FileNotFoundError, NotADirectoryError):
+                # One of the path components was not a directory, likely
+                # because we're in a zip file.
+                write = False
+            except PermissionError:
```

Comment:
> exception PermissionError
Raised when trying to run an operation without the adequate access rights - for example filesystem permissions. Corresponds to errno EACCES and EPERM.

I couldn't find an exception corresponding to `errno.EROFS` (Read-only file system), but I suppose it will fall under the umbrella of `PermissionError`.